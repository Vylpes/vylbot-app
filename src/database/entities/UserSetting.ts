import { Column, Entity} from "typeorm";
import AppDataSource from "../dataSources/appDataSource";
import BaseEntity from "../../contracts/BaseEntity";

@Entity()
export default class UserSetting extends BaseEntity {
    constructor(userId: string, key: string, value: string) {
        super();

        this.UserId = userId;
        this.Key = key;
        this.Value = value;
    }

    @Column()
    UserId: string;

    @Column()
    Key: string;

    @Column()
    Value: string;

    public UpdateValue(value: string) {
        this.Value = value;
    }

    public static async FetchOneByKey(userId: string, key: string, relations?: string[]): Promise<UserSetting | null> {
        const repository = AppDataSource.getRepository(UserSetting);

        const single = await repository.findOne({ where: { UserId: userId, Key: key }, relations: relations || {} });

        return single;
    }
}
