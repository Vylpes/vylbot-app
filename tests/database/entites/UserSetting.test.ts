import AppDataSource from "../../../src/database/dataSources/appDataSource";
import UserSetting from "../../../src/database/entities/UserSetting";

describe("constructor", () => {
    let userSetting: UserSetting;

    beforeEach(() => {
        userSetting = new UserSetting("userId", "key", "value");
    });

    test("EXPECT settings to be configured", () => {
        expect(userSetting).toMatchSnapshot({
            Id: expect.any(String),
            WhenCreated: expect.any(Date),
            WhenUpdated: expect.any(Date),
        });
    });
});

describe("UpdateValue", () => {
    let userSetting: UserSetting;

    beforeEach(() => {
        userSetting = new UserSetting("userId", "key", "value");

        userSetting.UpdateValue("newValue");
    });

    test("EXPECT value to be updated", () => {
        expect(userSetting.Value).toBe("newValue");
    });
});

describe("FetchOneByKey", () => {
    let result: UserSetting | null;
    let userSetting: UserSetting;

    let findOneMock: jest.Mock;

    beforeEach(async () => {
        userSetting = new UserSetting("userId", "key", "value");

        findOneMock = jest.fn().mockResolvedValue(userSetting);

        AppDataSource.getRepository = jest.fn().mockReturnValue({
            findOne: findOneMock,
        });

        result = await UserSetting.FetchOneByKey("userId", "key");
    });

    test("EXPECT getRepository to have been called", () => {
        expect(AppDataSource.getRepository).toHaveBeenCalledTimes(1);
        expect(AppDataSource.getRepository).toHaveBeenCalledWith(UserSetting);
    });

    test("EXPECT repository.findOne to have been called", () => {
        expect(findOneMock).toHaveBeenCalledTimes(1);
        expect(findOneMock).toHaveBeenCalledWith({ where: { UserId: "userId", Key: "key" }, relations: {}});
    })

    test("EXPECT single entity returned", () => {
        expect(result).toBe(userSetting);
    });
});
