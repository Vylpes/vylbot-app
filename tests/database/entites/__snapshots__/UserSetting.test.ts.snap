// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor EXPECT settings to be configured 1`] = `
{
  "Id": Any<String>,
  "Key": "key",
  "UserId": "userId",
  "Value": "value",
  "WhenCreated": Any<Date>,
  "WhenUpdated": Any<Date>,
}
`;
